lib/Support/CompilationLog.cpp

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Support/CompilationLog.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <fstream>
+
+namespace glow {
+static llvm::cl::opt<bool>
+    dumpCompilationLogOpt("dump-compilation-log",
+                          llvm::cl::desc("Dump compilation log"));
+
+CompilationScope::CompilationScope(llvm::StringRef scopeName)
+    : scopeName_(std::move(scopeName)) {
+  if (dumpCompilationLogOpt) {
+    std::string startScopeMsg("\n---Entering the compilation scope: ");
+    dumpCompilationLog(startScopeMsg.append(scopeName_));
+  }
+}
+
+CompilationScope::~CompilationScope() {
+  if (dumpCompilationLogOpt) {
+    std::string endScopeMsg("\n---Leaving the compilation scope: ");
+    dumpCompilationLog(endScopeMsg.append(scopeName_));
+  }
+}
+
+void dumpCompilationLog(llvm::StringRef logContent) {
+  dumpCompilationLog("compile.log", logContent);
```

Comment:
Consider either adding a comment to explain what `"compile.log"` is:
```
dumpCompilationLog(/*logFilename=*/"compile.log", logContent);
```
or define `"compile.log"` as a `constexpr` at the top of the file with a descriptive name:
```
namespace {
auto constexpr defaultCompileLogFilename = "compile.log";
} // namespace
...
dumpCompilationLog(defaultCompileLogFilename, logContent);
````